source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
-                    update_start = false;
-
-                    block_data_size = -block_data_size;
-                    if (block_data_size >= data_size) {
-                        /*found block*/
-                        if (block_data_size > (data_size + 4)) {
-                            //There is enough room for a new hole so create it first
-                            int *hole_ptr;
-                            if (direction > 0) {
-                                hole_ptr = ptr + data_size + 2;
-                            } else {
-                                hole_ptr = ptr - (data_size + 2);
-                            }
-                            int hole_size = block_data_size - data_size - 2;
-                            *hole_ptr = -(hole_size);
-                            if (direction > 0) {
-                                hole_ptr += (hole_size + 1);
+                int block_data_size = -(*ptr);
+
+                if (block_data_size >= data_size) {
+                    /*found block*/
+                    if (block_data_size >= (data_size + 2 + HOLE_T_SIZE)) {
+                        int hole_size = block_data_size - data_size - 2;
+
+                        //There is enough room for a new hole so create it first
+                        if ( direction > 0 ) {
+                            hole_ptr = ptr + data_size + 2;
+                        } else {
+                            hole_ptr = ptr;
+                        }
+                        if( direction > 0 ){
+                            // Would like to just replace this cur_hole with new descriptor, but
+                            // they could overlap, so ns_list_replace might fail
+                            //ns_list_replace(&holes_list, cur_hole, hole_from_block_start(hole_ptr));
+                            hole_t *before = ns_list_get_previous(&holes_list, cur_hole);
+                            ns_list_remove(&holes_list, cur_hole);
+                            if (before) {
+                                ns_list_add_after(&holes_list, before, hole_from_block_start(hole_ptr));
                             } else {
-                                hole_ptr -= (hole_size + 1);
+                                ns_list_add_to_start(&holes_list, hole_from_block_start(hole_ptr));
                             }
-                            *hole_ptr = -(hole_size);
-                        }else{
-                            data_size = block_data_size;
                         }
 
-                        *ptr = data_size;
-                        if (direction > 0) {
-                            ptr++;
-                            retval = ptr;
-                            ptr += data_size;
-                        } else {
-                            ptr -= data_size;
-                            retval = ptr;
-                            ptr--;
+                        *hole_ptr = -(hole_size);
```

Comment:
Simplify to:

           hole_ptr[0] = -hole_size;
           hole_ptr[1 + hole_size] = -hole_size;
           if (direction < 0) {
                 ptr += 1 + hole_size + 1;  // I like this better than +2 cos it shows the actual layout.
           }